src/core/or/sendme.c

```diff
@@ -571,31 +594,25 @@ sendme_note_stream_data_packaged(edge_connection_t *conn)
   return --conn->package_window;
 }
 
-/* Note the cell digest in the circuit sendme last digests FIFO if applicable.
- * It is safe to pass a circuit that isn't meant to track those digests. */
+/* Record the cell digest into the circuit sendme digest list depending on
+ * which edge we are. The digest is recorded only if we expect the next cell
+ * that we will receive is a SENDME so we can match the digest. */
 void
-sendme_record_cell_digest(circuit_t *circ)
+sendme_record_cell_digest_on_circ(circuit_t *circ, crypt_path_t *cpath)
 {
-  const uint8_t *digest;
+  int package_window;
+  uint8_t *sendme_digest;
 
   tor_assert(circ);
 
-  /* We only keep the cell digest if we are the Exit on that circuit and if
-   * this cell is the last one before the client should send a SENDME. */
-  if (CIRCUIT_IS_ORIGIN(circ)) {
-    return;
-  }
-  /* Is this the last cell before a SENDME? The idea is that if the
-   * package_window reaches a multiple of the increment, after this cell, we
-   * should expect a SENDME. */
-  if (!sendme_circuit_cell_is_next(circ->package_window)) {
-    return;
+  if (cpath) {
+    package_window = cpath->package_window;
+    sendme_digest = cpath_get_sendme_digest(cpath);
```

Comment:
This looks like we're moving backwards here: like we are computing the sendme digest even when we are not going to record it. Is that correct?  Remember, these digests are not cheap to calculate.

(edited to add) -- but it looks like we add other conditions later?  I'm a bit confused, I'll have to come back to this branch.

Reply:
Indeed... Very good catch.

I've fix this in the fixup commit f78fb4349da19daf and added a comment to this part about the CPU cost.